### PR TITLE
add init container for updating ca trust and shift getting ca cert from secret to config map

### PR DESCRIPTION
**Pull request by mittal-ishaan**
## What does this PR change?
* Create a init container to update ca trust which we earlier expected the user to do inside the container itself.
Taken reference from https://jfrog.com/help/r/artifactory-how-to-add-custom-ssl-certificates-to-an-artifactory-pod-using-helm-charts/artifactory-how-to-add-custom-ssl-certificates-to-an-artifactory-pod-using-helm-charts

## Does this PR rely on any other PRs?
https://github.com/kubecost/cost-analyzer-helm-chart/pull/3760

## How does this PR impact users? (This is the kind of thing that goes in release notes!)
Users will now not need to run update-ca-trust command inside the container. The init container will do it.

## How was this PR tested?
Deployed the kubecost deployment with `caCertsSecret` helm value set and ca cert as Secret created.
reading through `/etc/pki/ca-trust/extracted/openssl/ca-bundle.trust.crt` we see it updated with the given ca cert in Secret.

helm values:

```yaml
global:
  updateCaTrust:
    enabled: false  # Set to true to enable the init container for updating CA trust
    # Security context settings for the init container.
    securityContext:
      runAsUser: 0
      runAsGroup: 0
      runAsNonRoot: false
      allowPrivilegeEscalation: false
      readOnlyRootFilesystem: true
      seccompProfile:
        type: RuntimeDefault
    caCertsSecret: ca-certs-secret  # The name of the Secret containing custom CA certificates to mount to the cost-model container.
    # caCertsConfig: ca-certs-config  # The name of the ConfigMap containing the CA trust configuration.
    resources: {}  # Resource requests and limits for the init container.
    caCertsMountPath: /etc/pki/ca-trust/source/anchors  # The path where the custom CA certificates will be mounted in the init container

```

### CODE REVIEW

File: cost-analyzer/templates/cost-analyzer-deployment-template.yaml [hunk 1]
```diff
@@ -351,6 +353,28 @@ spec:
           securityContext:
             runAsUser: 0
 {{ end }}
+        {{- if .Values.kubecostModel.caCertsSecret }}
+        - name: "sslsetup"
```
Comment (jessegoodier) [inline cost-analyzer/templates/cost-analyzer-deployment-template.yaml]:
```suggestion
        - name: "update-ca-trust"
```
what do you think of changing the container name?

File: cost-analyzer/templates/cost-analyzer-deployment-template.yaml [hunk 2]
```diff
@@ -351,6 +353,28 @@ spec:
...
+          image: {{ include "cost-model.image" . | trim | quote}}
+          imagePullPolicy: Always
+          securityContext:
```
Comment (jessegoodier) [inline cost-analyzer/templates/cost-analyzer-deployment-template.yaml]: can we template this and add these as the defaults in values.yaml?

Comment (jessegoodier) [reply cost-analyzer/templates/cost-analyzer-deployment-template.yaml]: thinking `kubecostModel.update-ca-trust.securityContext`

Comment (mittal-ishaan) [reply cost-analyzer/templates/cost-analyzer-deployment-template.yaml]: agree here

File: cost-analyzer/templates/cost-analyzer-deployment-template.yaml (hunk 1 above)
Comment (mittal-ishaan) [reply cost-analyzer/templates/cost-analyzer-deployment-template.yaml]: agree

File: cost-analyzer/templates/cost-analyzer-deployment-template.yaml [hunk 3]
```diff
@@ -351,6 +353,30 @@ spec:
           securityContext:
             runAsUser: 0
 {{ end }}
+        {{- if .Values.kubecostModel.updateCaTrust.enabled }}
+        - name: "update-ca-trust"
+          image: {{ include "cost-model.image" . | trim | quote}}
+          imagePullPolicy: Always
```
Comment (jessegoodier) [inline cost-analyzer/templates/cost-analyzer-deployment-template.yaml:367]:
```suggestion
          imagePullPolicy: IfNotPresent
```

File: cost-analyzer/templates/cost-analyzer-deployment-template.yaml [hunk 4]
```diff
@@ -351,6 +353,30 @@ spec:
           securityContext:
             runAsUser: 0
 {{ end }}
+        {{- if .Values.kubecostModel.updateCaTrust.enabled }}
+        - name: "update-ca-trust"
```
Comment (jessegoodier) [inline cost-analyzer/templates/cost-analyzer-deployment-template.yaml]:
```suggestion
        - name: update-ca-trust
```

**Comment by jessegoodier** [comment]
This is looking good to me. Tested with custom secret name.

Will ping customer and merge after we get feedback.

**Comment by chipzoller** [comment]
But why a secret? Root certificates aren't sensitive.

File: cost-analyzer/values.yaml [hunk 5]
```diff
@@ -575,8 +575,19 @@ kubecostModel:
...
 
-  # the name of the Secret containing custom CA certs to mount to cost model container
-  # caCertsSecret: ca-certs-secret
+  # Install the cost-model with a custom CA certs.
+  # The following configuration is for an init container to update the CA trust.
+  updateCaTrust:
+    enabled: false  # Set to true to enable the init container for updating CA trust
+    # Security context settings for the init container.
+    securityContext:
+      privileged: true
```
Comment (jessegoodier) [inline cost-analyzer/values.yaml]:
```suggestion
```

Comment (jessegoodier) [reply cost-analyzer/values.yaml]: looks like it works without this line?

**Comment by mittal-ishaan** [comment]
@chipzoller The user requested it to be a secret as they store/retrieve certs as secrets internally.

**Comment by chipzoller** [comment]
Then maybe make this so it accepts a Secret or ConfigMap?

**Comment by mittal-ishaan** [comment]
sure, we should do that. let me get that in.

**Comment by thomasvn** [comment]
Future TODO. This updates the CA trust and certificates for the cost-analyzer pod. But it's likely that users would need to perform the same updates across other pods like the aggregator and cloudcost pod.

**Comment by jessegoodier** [comment]
@mittal-ishaan great work here. Thanks for all of the iterations while we figured out the best approach.

Agree with Thomas that we should add to aggregator and cloud cost when you have a moment.

I'm now questioning if we should have put this under a more general section instead of nesting under `kubecostModel` ?

@chipzoller and @thomasvn what are your thoughts?

**Comment by thomasvn** [comment]
Yes! I like the idea of doing `.Values.updateCaTrust` instead of `.Values.kubecostModel.updateCaTrust`.

**Comment by chipzoller** [comment]
Also agree that I'd try and pull this out and generalize it, but I'm not sure the root level is the best place.

You guys have any other suggestions for where these values may best go?

**Comment by thomasvn** [comment]
Only other place that seems suitable is under `.Values.global`. I'm open to other ideas though?

**Comment by jessegoodier** [comment]
> Only other place that seems suitable is under `.Values.global`. I'm open to other ideas though?

I'm thinking `global.` is better than `productConfigs`

File: cost-analyzer/values.yaml [hunk 6]
```diff
@@ -226,6 +226,24 @@ global:
       drop:
       - ALL
 
+  # Install the cost-model with a custom CA certs.
+  # The following configuration is for an init container to update the CA trust.
```
Comment (thomasvn) [inline cost-analyzer/values.yaml]:
```suggestion
  # Installs custom CA certificates onto Kubecost pods
```

**Review by thomasvn — APPROVED** [review]
This looks good to me. These configs are now available under `.Values.global.updateCaTrust`.

Approving for immediate merge so fixes are available to customer. However we now need a follow-up PR to make similar changes to the Kubecost Aggregator and Kubecost CloudCost pods, then test those changes.

**Comment by jessegoodier** [comment]
Thanks Thomas. great work here team.

**Comment by srpomeroy** [comment]
Did we ever add docs for this feature?

**Comment by thomasvn** [comment]
@srpomeroy Thanks for following up. I'm not sure we ever did.

@mittal-ishaan Could you please create a Docs PR to describe how to configure `.Values.global.updateCaTrust`? It may end up being a new markdown page under the "Advanced Configuration" section [here](https://github.com/kubecost/docs/blob/main/install-and-configure/advanced-configuration/README.md). Alternatively, you could add more robust comments about how to configure this directly into the `values.yaml` file.

**Pull request merged**